### PR TITLE
Adding total-limit-time option to acsf-tools-ml command.

### LINKED ISSUE #44: Add a --total-time-limit option to acsf-tools-ml command
**Context**
I need to process a queue across an ACSF instance with several (80+) sites. Ideally I would like a way to execute as many items as possible across sites on a given time limit so the servers aren't overwhelmed by queue processing for a very long time.

**Propposed solution**
Allow acsf-tools users to specify a --total-time-limit option in the acsf-tools-ml command. If this option is present, the given command will be executed multiple times within the given time limit.

### PR DESCRIPTION
**Pull request by jose-oliveira**
Fixes #44.

### CODE REVIEW

**Comment by jose-oliveira** [comment]
@sdelbosc Nope, the goal is to run them sequentially.

> I suspect that you alter the behaviour of current commands 
Can you expand on that?

**Comment by sdelbosc** [comment]
Second try :-)
After having read your last comment, I have re-reviewed your changes and I invalidate my comment about possible impact. Looks like that if you don't use that option value would be zero and impact on current commands would be null (apart maybe from some notices).
However, I am definitely not clear on what this code is changing as it seems that the time end is evaluated after all the commands have been triggered.
We should probably have a quick call.

**Comment by sdelbosc** [comment]
Just out of a chat with Jose.
I understand better what he wants to achieve with this PR.

Basically allocate an amount of time for a drush command and repeat this command over the sites of the factory until the time is exhausted. If all the sites have been processed before the end of the allocated time, acsf-tools-ml will loop over the sites again.

File: acsf_tools.drush.inc [hunk 1]
```diff
@@ -179,46 +179,56 @@ function drush_acsf_tools_ml() {
...
-      // find the settings file, we use drush core-status to get it.
-      if (!empty($profiles)) {
-        $proceed = TRUE;
-
-        $site_settings_filepath = 'sites/g/files/' . $details['name'] . '/settings.php';
-        if (file_exists($site_settings_filepath)) {
-          $site_settings = @file_get_contents($site_settings_filepath);
-          if (preg_match("/'install_profile'] = '([a-zA-Z_]*)'/", $site_settings, $matches)) {
-            if (isset($matches[1]) && !in_array($matches[1], $profiles)) {
+    $total_time_limit = (int) drush_get_option('total-time-limit');
```
Comment (sdelbosc) [inline acsf_tools.drush.inc]: We should update drush help to mention this parameter.

Comment (sdelbosc) [reply acsf_tools.drush.inc]: We should also have a default value of 0.

File: acsf_tools.drush.inc [hunk 2]
```diff
@@ -179,46 +179,56 @@ function drush_acsf_tools_ml() {
...
           }
         }
 
-        if (!$proceed) {
-          drush_log("\n=> Skipping command on $domain", 'ok');
-          continue;
+        drush_log("\n=> Running command on $domain", 'ok');
+        $output = drush_invoke_process('@self', $command, $arguments, $options + array('l' => $domain));
+
+        // @TODO Find a better way to remove sites from execution list.
```
Comment (sdelbosc) [inline acsf_tools.drush.inc]: This is likely not generic enough given that we could run any drush command.
I would rather add an optional parameter called time-limit-site-removal-condition that would contain a regexp. If drush output matches that regexp then the site is removed from the list of sites.

Comment (sdelbosc) [reply acsf_tools.drush.inc:226]: @jose-oliveira how essential is this in the implementation? This looks to me like a nice to have improvement for your very specific use case. So I am tempted to remove from this PR.

**Comment by mirsoftacquia** [comment]
Please close in favour of #68 which contains fixes in comments here.

**Comment by sdelbosc** [comment]
Supported for drush 9 only
See https://github.com/acquia/acsf-tools/pull/69

**Pull request closed without merging**